File: plugins/pytest-agent-observability/src/pytest_agent_observability/events.py
"""Serialize LiveKit RunResult events into the plain-dict form the server expects."""

from __future__ import annotations

import json
from typing import Any, Iterable, Optional


# Field caps (mirrors the server's payload-size guidance).
MAX_CONTENT_CHARS = 10_000
MAX_EVENTS_PER_CASE = 500


def serialize_events(run_events: Optional[Iterable[Any]]) -> list[dict]:
    """Turn RunEvent dataclass instances into JSON-ready dicts.

    Returns an empty list when `run_events` is None/empty. Unknown event types
    are skipped silently rather than raising — the plugin must not break tests.
    """
    if not run_events:
        return []
    out: list[dict] = []
    for ev in run_events:
        d = _serialize_event(ev)
        if d is None:
            continue
        out.append(d)
        if len(out) >= MAX_EVENTS_PER_CASE:
            break
    return out


def _serialize_event(ev: Any) -> Optional[dict]:
    ev_type = getattr(ev, "type", None)
    if ev_type == "message":
        item = getattr(ev, "item", None)
        role = _attr(item, "role")
        # LiveKit ChatMessage exposes text_content as a convenience property.
        content = _attr(item, "text_content")
        if content is None:
            content = _attr(item, "content")
            if isinstance(content, list):
                content = "".join(str(c) for c in content)
        return _trim_content({
            "type": "message",
            "role": role,
            "content": content,
            "interrupted": bool(_attr(item, "interrupted", default=False)),
        })

    if ev_type == "function_call":
        item = getattr(ev, "item", None)
        raw_args = _attr(item, "arguments")
        args = raw_args
        if isinstance(raw_args, str):
            try:
                args = json.loads(raw_args)
            except Exception:
                args = raw_args  # keep as string if not valid JSON
        return {
            "type": "function_call",
            "name": _attr(item, "name"),
            "arguments": args,
            "call_id": _attr(item, "call_id") or _attr(item, "id"),
        }

    if ev_type == "function_call_output":
        item = getattr(ev, "item", None)
        return _trim_output({
            "type": "function_call_output",
            "output": _attr(item, "output"),
            "is_error": bool(_attr(item, "is_error", default=False)),
            "call_id": _attr(item, "call_id"),
        })

    if ev_type == "agent_handoff":
        old_agent = _attr(ev, "old_agent")
        new_agent = _attr(ev, "new_agent")
        return {
            "type": "agent_handoff",
            "from_agent": _class_name(old_agent),
            "to_agent": _class_name(new_agent),
        }

    return None


def _attr(obj: Any, name: str, *, default: Any = None) -> Any:
    if obj is None:
        return default
    return getattr(obj, name, default)


def _class_name(obj: Any) -> Optional[str]:
    if obj is None:
        return None
    cls = getattr(obj, "__class__", None)
    return cls.__name__ if cls is not None else None


def _trim_content(d: dict) -> dict:
    c = d.get("content")
    if isinstance(c, str) and len(c) > MAX_CONTENT_CHARS:
        d["content"] = c[:MAX_CONTENT_CHARS] + "…"
    return d


def _trim_output(d: dict) -> dict:
    o = d.get("output")
    if isinstance(o, str) and len(o) > MAX_CONTENT_CHARS:
        d["output"] = o[:MAX_CONTENT_CHARS] + "…"
    return d

File: plugins/agent-observability-sdk-node/src/livekit/vitest/events.ts
import type { RunEvent } from "./types.js";

/**
 * Serialize LiveKit RunResult events into JSON-ready dicts.
 *
 * Philosophy: do the *minimum* transformation needed to make the event
 * JSON-serializable, and pass everything else through untouched. No field
 * hand-picking, no truncation, no silent drops of unknown event types.
 *
 * The only transforms we apply:
 *  - Convert the event (and its nested `item`) to a plain dict, preserving
 *    every field — including `item.metrics`, timestamps, IDs, etc.
 *  - Recursively snake_case object keys so the wire format matches the
 *    Python plugin's Pydantic projection. The obs dashboard reads
 *    snake_case (`llm_node_ttft`, `created_at`, `is_error`, …) and the
 *    Node SDK ships native camelCase, so without this normalizer
 *    vitest payloads would render with empty metric chips and
 *    inconsistent item fields.
 *  - For `function_call`, parse `arguments` from JSON string → object as a
 *    convenience (so the dashboard doesn't have to).
 *  - For `agent_handoff`, replace the `old_agent` / `new_agent` instance
 *    references with their constructor names (full Agent objects aren't
 *    JSON-friendly).
 */
export function serializeEvents(rawEvents: unknown[] | null | undefined): RunEvent[] {
  if (!rawEvents || rawEvents.length === 0) return [];
  const out: RunEvent[] = [];
  for (const ev of rawEvents) {
    try {
      out.push(snakeifyKeys(serializeEvent(ev)) as RunEvent);
    } catch {
      out.push({
        type: String((ev as any)?.type ?? "unknown"),
        _serialize_error: true,
        repr: String(ev),
      } as unknown as RunEvent);
    }
  }
  return out;
}

function camelToSnake(key: string): string {
  return key.replace(/[A-Z]/g, (c) => `_${c.toLowerCase()}`);
}

/**
 * Recursively rewrite object keys from camelCase to snake_case. Arrays and
 * primitives pass through untouched. Idempotent on already-snake_case input
 * (no `[A-Z]` chars, so the regex is a no-op).
 */
function snakeifyKeys(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(snakeifyKeys);
  if (value && typeof value === "object") {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      out[camelToSnake(k)] = snakeifyKeys(v);
    }
    return out;
  }
  return value;
}

function serializeEvent(ev: unknown): RunEvent {
  const ev_dict = toPlainDict(ev);
  const type = (ev_dict as any).type ?? (ev as any)?.type;

  const rawItem = (ev as any)?.item;
  if (rawItem !== undefined) {
    (ev_dict as any).item = toPlainDict(rawItem);
  }
  const item = (ev_dict as any).item ?? {};

  if (type === "message") {
    let content = item.textContent ?? item.text_content ?? item.content;
    if (Array.isArray(content)) content = content.map(String).join("");
    setDefault(ev_dict, "role", item.role);
    setDefault(ev_dict, "content", typeof content === "string" ? content : undefined);
    setDefault(ev_dict, "interrupted", Boolean(item.interrupted));
    // Pass per-turn metrics through so the dashboard can render latency.
    if ("metrics" in item) setDefault(ev_dict, "metrics", item.metrics);
  } else if (type === "function_call") {
    // LiveKit Node's FunctionCall chat item stores the JSON-string args
    // under `args`, not `arguments`. Python's FunctionCall uses
    // `arguments`. Accept both so either SDK flavor produces a populated
    // payload instead of `arguments: undefined`.
    let args: unknown = item.arguments ?? item.args;
    if (typeof args === "string") {
      try {
        args = JSON.parse(args);
      } catch {
        /* keep original string */
      }
    }
    setDefault(ev_dict, "name", item.name);
    setDefault(ev_dict, "arguments", args);
    setDefault(ev_dict, "call_id", item.call_id ?? item.callId ?? item.id);
  } else if (type === "function_call_output") {
    setDefault(ev_dict, "output", item.output);
    setDefault(ev_dict, "is_error", Boolean(item.is_error ?? item.isError));
    setDefault(ev_dict, "call_id", item.call_id ?? item.callId);
  } else if (type === "agent_handoff") {
    const oldAgent =
      (ev as any)?.old_agent ?? (ev as any)?.oldAgent;
    const newAgent =
      (ev as any)?.new_agent ?? (ev as any)?.newAgent;
    (ev_dict as any).from_agent = className(oldAgent);
    (ev_dict as any).to_agent = className(newAgent);
    delete (ev_dict as any).old_agent;
    delete (ev_dict as any).oldAgent;
    delete (ev_dict as any).new_agent;
    delete (ev_dict as any).newAgent;
  }

  return ev_dict as unknown as RunEvent;
}

/** Best-effort conversion of any object to a plain JSON-friendly dict. */
function toPlainDict(obj: unknown): Record<string, unknown> {
  if (obj == null || typeof obj !== "object") return {};
  if (Array.isArray(obj)) return { ...obj };
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(obj)) {
    if (k.startsWith("_")) continue;
    if (typeof v === "function") continue;
    out[k] = v;
  }
  return out;
}

function setDefault(
  target: Record<string, unknown>,
  key: string,
  value: unknown,
): void {
  if (!(key in target)) target[key] = value;
}

function className(obj: unknown): string | undefined {
  if (!obj) return undefined;
  const ctor = (obj as { constructor?: { name?: string } }).constructor;
  return ctor?.name;
}
